plugins/blockvault_client_plugin/postgres_backend.cpp

```diff
@@ -0,0 +1,196 @@
+#include "postgres_backend.hpp"
+#include <boost/filesystem.hpp>
+#include <fstream>
+#include <iostream>
+#include <limits.h>
+#include <unistd.h>
+
+namespace eosio {
+namespace blockvault {
+
+postgres_backend::postgres_backend(const std::string& options)
+    : conn(options) {
+   pqxx::work w(conn);
+   w.exec("CREATE TABLE IF NOT EXISTS BlockData (watermark_bn bigint, watermark_ts bigint, lib bigint, "
+          "block_id bytea, previous_block_id bytea, block oid, block_size bigint);"
+          "CREATE TABLE IF NOT EXISTS SnapshotData (watermark_bn bigint, watermark_ts bigint, snapshot oid);");
+   w.commit();
+
+   conn.prepare(
+       "insert_constructed_block",
+       "INSERT INTO BlockData (watermark_bn, watermark_ts, lib, block_id, previous_block_id, block, block_size) "
+       "SELECT $1, $2, $3, $4, $5, $6, $7  WHERE NOT "
+       "EXISTS (SELECT * FROM BlockData WHERE (watermark_bn >= $1) OR (watermark_ts > $2) OR (lib > $3) OR (block_id "
```

Comment:
> if my read is correct, I think this should be `(watermark_ts >= $2)`. It is considered "double production" to produce two blocks at the same height OR the same timestamp. This statement seems to require that there is no known block with a higher or equal blocknumber OR higher timestamp which leaves open a higher blocknumber with the same timstamp.

The timestamp only changes when producer changes, i.e. it is only incremented every 12 blocks in the ideal cases. If you have the (watermark_ts >= $2), that means the producer can only produce one block every production round.

Reply:
The timestamp should change with every block.  If it isn't that is a bug.  That time should be the blocks' timestamps.  

### recording watermarks
this code is fired from the `accepted_block_header` signal from controller:

https://github.com/EOSIO/eos/blob/efadbb1/plugins/producer_plugin/producer_plugin.cpp#L229-L231

That signal is fired from `push_block` for every block that is accepted to fork-db (has proper headers) and from `commit_block` for locally constructed blocks.  That should result in it being updated every block. 


### use
This code is what prevents double signing on a node that has complete data:

https://github.com/EOSIO/eos/blob/efadbb1/plugins/producer_plugin/producer_plugin.cpp#L1374-L1389

It is a little convoluted because it is checks `>` the block num of the *current head block* (which will be our previous block making our block num 1 higher) AND it checks that our candidate `block_timestamp` is `>=` the watermarks timestamp.

Reply:
fixed in commit 900909f2acc93a6b7b7ccfbeb5d35b7eab628b82